backend/internal/service/handler/auth.go

```diff
@@ -0,0 +1,155 @@
+	CodeVerifierStore *CodeVerifierStore
+}
+
+func NewAuthHandler() *AuthHandler {
+	return &AuthHandler{
+		CodeVerifierStore: &CodeVerifierStore{
+			CodeVerifierMap: make(map[string]string),
+		},
+	}
+}
+
+func (h *AuthHandler) SpotifyRedirect(c *fiber.Ctx) error {
+	verifier := oauth2.GenerateVerifier()
+	challenge := oauth2.S256ChallengeFromVerifier(verifier)
+
+	// store state as uuid for now
+	state := uuid.NewString()
+
+	ctx, cancel := context.WithTimeout(context.TODO(), 1*time.Minute)
+
+	h.CodeVerifierStore.Mu.Lock()
+	h.CodeVerifierStore.CodeVerifierMap[state] = verifier
+	h.CodeVerifierStore.Mu.Unlock()
+
+	go func() {
+		<-ctx.Done()
+		h.CodeVerifierStore.Mu.Lock()
+		delete(h.CodeVerifierStore.CodeVerifierMap, state)
+		h.CodeVerifierStore.Mu.Unlock()
+		cancel()
+		fmt.Printf("state %s has been deleted\n", state)
+	}()
+
+	params := url.Values{}
+	params.Add("client_id", os.Getenv("SPOTIFY_CLIENT_ID"))
+	params.Add("response_type", "code")
+	params.Add("redirect_uri", os.Getenv("SPOTIFY_REDIRECT_URI"))
+	params.Add("state", state)
+	params.Add("scope", "user-read-private")
+	params.Add("code_challenge_method", "S256")
+	params.Add("code_challenge", challenge)
+
+	u, err := url.Parse(SpotifyAuthURL)
+	if err != nil {
+		return errs.InternalServerError()
+	}
+
+	u.RawQuery = params.Encode()
+
+	// remove in future
+	fmt.Printf("%s\n", u.String())
+
+	return c.Redirect(u.String())
+}
+
+func (h *AuthHandler) SpotifyCallback(c *fiber.Ctx) error {
+	if err := c.Query("error"); err != "" {
+		// indicates that user denied access
+		return errs.Unauthorized()
+	}
+
+	code := c.Query("code")
+	state := c.Query("state")
+
+	verifier, ok := h.CodeVerifierStore.CodeVerifierMap[state]
+	if !ok {
+		// indicates that state is different from what was originally stored
+		return errs.Unauthorized()
+	}
+
+	params := url.Values{}
+	params.Add("grant_type", "authorization_code")
+	params.Add("code", code)
+	params.Add("redirect_uri", os.Getenv("SPOTIFY_REDIRECT_URI"))
+	params.Add("client_id", os.Getenv("SPOTIFY_CLIENT_ID"))
+	params.Add("code_verifier", verifier)
+
+	req, err := http.NewRequest("POST", SpotifyTokenURL, nil)
+	if err != nil {
+		return errs.InternalServerError()
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.URL.RawQuery = params.Encode()
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return errs.InternalServerError()
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return errs.NewHTTPError(500, errors.New("failed to get token"))
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return errs.InternalServerError()
+	}
+
+	var tokenResponse TokenResponse
+	if err := json.Unmarshal(body, &tokenResponse); err != nil {
+		return errs.InternalServerError()
+	}
+
+	// remove in future
+	fmt.Printf("%+v\n", tokenResponse)
+
+	// do something with tokenResponse
+
+	// redirect to frontend
+	return c.Redirect("http://localhost:3000")
```

Comment:
can you get away with
```suggestion
	return c.Redirect("http://127.0.0.1:3000")
```
when using `localhost`, your system may first attempt to resolve it to both IPv4 and IPv6 addresses, which involves a DNS lookup process